File: src/go/problems/AddTwoNumbers_test.go
// https://leetcode.com/problems/add-two-numbers/

package problems

import "testing"

func TestAddTwoNumbers(t *testing.T) {
	l1 := ListNode{
		Val: 2,
		Next: &ListNode{
			Val: 4,
			Next: &ListNode{
				Val: 3,
			},
		},
	}

	var l2 = ListNode{
		Val: 5,
		Next: &ListNode{
			Val: 6,
			Next: &ListNode{
				Val: 4,
			},
		},
	}

	result := addTwoNumbers(&l1, &l2)
	t.Log(result)
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	node1 := l1
	node2 := l2
	result := &ListNode{}
	current := result
	var carry = 0

	for node1 != nil || node2 != nil || carry > 0 {
		var sum = carry
		if node1 != nil {
			sum = sum + node1.Val
			node1 = node1.Next
		}

		if node2 != nil {
			sum = sum + node2.Val
			node2 = node2.Next
		}

		carry = sum / 10
		current.Next = &ListNode{Val: sum % 10}
		current = current.Next
	}

	return result.Next
}

File: src/python/problems/add_two_numbers_test.py
# https://leetcode.com/problems/add-two-numbers/

from typing import Optional

from models.list_node import *
from core.problem_base import *

class AddTwoNumbers(ProblemBase):
    def Solution(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        node1 = l1
        node2 = l2
        result = ListNode()
        current = result
        carry = 0

        while node1 or node2 or carry > 0:
            num1 = node1.val if node1 else 0
            num2 = node2.val if node2 else 0
            sum = carry + num1 + num2
            carry = sum // 10
            current.next = ListNode(sum % 10)
        
            current = current.next
            node1 = node1.next if node1 else None
            node2 = node2.next if node2 else None

        return result.next

if __name__ == '__main__':
    TestGen(AddTwoNumbers) \
        .Add(lambda tc: tc.ParamListNode([2,4,3]).ParamListNode([5,6,4]).ResultListNode([7,0,8])) \
        .Add(lambda tc: tc.ParamListNode([0]).ParamListNode([0]).ResultListNode([0])) \
        .Add(lambda tc: tc.ParamListNode([9,9,9,9,9,9,9]).ParamListNode([9,9,9,9]).ResultListNode([8,9,9,9,0,0,0,1])) \
        .Run()